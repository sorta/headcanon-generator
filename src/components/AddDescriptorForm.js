import React, { Component } from 'react';
import PropTypes from 'prop-types';

class AddDescriptorForm extends Component {
  static propTypes = {
    addDescriptor: PropTypes.func,
  };

  static defaultProps = {
    addDescriptor: () => {},
  };

  nameRef = React.createRef();

  createDescriptor = (ev) => {
    ev.preventDefault();
    const descriptor = {
      name: this.nameRef.current.value,
    };
    this.props.addDescriptor(descriptor);
    ev.currentTarget.reset();
  }

  render() {
    return (
      <form className="AddDescriptorForm form-add" onSubmit={this.createDescriptor}>
        <input name="name" ref={this.nameRef} type="text" placeholder="Name" required />
        <button type="submit">+</button>
      </form>
    );
  }
}

export default AddDescriptorForm;
