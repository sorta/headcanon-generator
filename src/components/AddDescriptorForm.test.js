import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import AddDescriptorForm from './AddDescriptorForm';

const props = {
  addDescriptor: jest.fn(),
  isManaging: () => false,
};

// Tests for regular mode
describe("AddDescriptorForm", () => {
  const component = shallow(<AddDescriptorForm addDescriptor={props.addDescriptor} isManaging={props.isManaging} />);

  it('renders and matches our snapshot', () => {
    const snapshot = renderer.create(
      <AddDescriptorForm addDescriptor={props.addDescriptor} isManaging={props.isManaging} />
    );
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should be null when not managing', () => {
    expect(component.getElement()).toBe(null);
  });
});


// Tests for managing mode
props.isManaging = () => true;
describe("AddDescriptorForm in managing mode", () => {
  // const component = shallow(<AddDescriptorForm addDescriptor={props.addDescriptor} isManaging={props.isManaging} />);

  it('renders and matches our snapshot', () => {
    const snapshot = renderer.create(
      <AddDescriptorForm addDescriptor={props.addDescriptor} isManaging={props.isManaging} />
    );
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have a name input field', () => {
    const component = shallow(<AddDescriptorForm addDescriptor={props.addDescriptor} isManaging={props.isManaging} />);
    expect(component.find('input[name="name"]').exists()).toBe(true);
  });

  it('should have a submit button', () => {
    const component = shallow(<AddDescriptorForm addDescriptor={props.addDescriptor} isManaging={props.isManaging} />);
    expect(component.find('button[type="submit"]').exists()).toBe(true);
  });

  it('form submit should call addDescriptor', () => {
    const component = mount(<AddDescriptorForm addDescriptor={props.addDescriptor} isManaging={props.isManaging} />);
    component.find('input[name="name"]').instance().value = 'TESTING';
    component.find('button[type="submit"]').simulate('submit');
    expect(props.addDescriptor.mock.calls.length).toEqual(1);
  });
});
