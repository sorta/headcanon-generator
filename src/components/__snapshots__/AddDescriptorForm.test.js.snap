// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddDescriptorForm in managing mode renders and matches our snapshot 1`] = `
<form
  className="AddDescriptorForm form-add control-row"
  onSubmit={[Function]}
>
  <input
    className="control-stretch"
    name="name"
    placeholder="New Descriptor Name"
    required={true}
    type="text"
  />
  <button
    type="submit"
  >
    +
  </button>
</form>
`;

exports[`AddDescriptorForm renders and matches our snapshot 1`] = `
<form
  className="AddDescriptorForm form-add control-row"
  onSubmit={[Function]}
>
  <input
    className="control-stretch"
    name="name"
    placeholder="New Descriptor Name"
    required={true}
    type="text"
  />
  <button
    type="submit"
  >
    +
  </button>
</form>
`;
